### PR TITLE
Allow GDExtension to register unexposed classes.

### PR DESCRIPTION
**Pull request by Daylily-Zeleen**
Because of renaming from GDNative to GDExtension, I need to remake [#69653](https://github.com/godotengine/godot/pull/69653). Otherwise I can't keep 1 commit.

For [#970](https://github.com/godotengine/godot-cpp/pull/970).

### CODE REVIEW

**Comment by akien-mga** [comment]
You should have been able to rebase this and the other PR, see https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html for details. But now that you opened new PRs it's fine, just something to keep in mind for later.

Both new PRs have unsuitable commit messages, "remake" is not informative. The commit message should say how the PR affects the engine, not what you did.

**Comment by Daylily-Zeleen** [comment]
> You should have been able to rebase this and the other PR, see https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html for details. But now that you opened new PRs it's fine, just something to keep in mind for later.
> 
> Both new PRs have unsuitable commit messages, "remake" is not informative. The commit message should say how the PR affects the engine, not what you did.

Thanks for your warning. I fixed the commit message.

**Comment by Zylann** [comment]
Does this actually prevent the class from being added to the Add X dialogs and Editor Help?
I found the latter to cause crashes on startup: https://github.com/godotengine/godot-cpp/issues/1179
Skipping such classes in these areas of the editor would fix that issue too.

**Comment by Daylily-Zeleen** [comment]
> Does this actually prevent the class from being added to the Add X dialogs and Editor Help?

The `exposed` menber will be used to filter document in DocTool, and prevent adding unexposed classes to `EditorHelp`, and I had tested it.
But the it seem it is not used in `CreateDialog`, so I'm not sure it will block unexposed classes to be added to `CreateDialog` or not, I have not tested for this.
If you have time, please do some test.

> I found the latter to cause crashes on startup, because Editor Help queries properties default values, which then instantiate those classes temporarily, and their constructor can want to access singletons that are not available: [godotengine/godot-cpp#1179](https://github.com/godotengine/godot-cpp/issues/1179)

You can avoid temporarily instantiation by registering abstract class. This is not the issue of this pr.

**Comment by dsnopek** [comment]
Now that PR https://github.com/godotengine/godot/pull/78634 has been merged, the `is_exposed` property could be moved to the new `GDExtensionClassCreationInfo2` struct, which would fix the compatibility issues from my previous comment.

@Daylily-Zeleen Is this a change you have time to do? If not, I'd be happy to take this over in a new PR :-)

**Comment by Daylily-Zeleen** [comment]
@dsnopek OK, now I moved it to `GDExtensionClassCreationInfo2`, both in [godot-cpp/pull/970](https://github.com/godotengine/godot-cpp/pull/970), too.

But I can't do a appropriate test about this, the editor helper (F1) can't show classes made by GDExtension currently maybe realat about [#76796](https://github.com/godotengine/godot/issues/76796), I'm not sure if it's working correctly, I just create a editor script like this:
```
@tool
extends EditorScript

func _run() -> void:
    print(ClassDB.class_existes("UnexposedClass"))
    print(ClassDB.class_existes("ExposedClass"))

```
And the Output panel show the reasult:
```
false
true
```

File: core/object/class_db.cpp [hunk 1]
```diff
@@ -1674,7 +1674,14 @@ void ClassDB::register_extension_class(ObjectGDExtension *p_extension) {
 	c.inherits = parent->name;
 	c.class_ptr = parent->class_ptr;
 	c.inherits_ptr = parent;
-	c.exposed = true;
+	c.exposed = p_extension->is_exposed;
+	if (c.exposed) {
+		// The parent classes should be exposed if has a exposed child class.
```
Comment (AThousandShips) [inline core/object/class_db.cpp:1679]:
```suggestion
		// The parent classes should be exposed if it has an exposed child class.
```
Grammar nitpick

File: core/object/class_db.h [hunk 2]
```diff
@@ -185,33 +185,61 @@ class ClassDB {
...
+		t->exposed = p_exposed;
 		t->class_ptr = T::get_class_ptr_static();
 		t->api = current_api;
 		//nothing
 	}
 
+	template <class T>
+	static void register_virtual_class(bool p_exposed = true) {
+		GLOBAL_LOCK_FUNCTION;
+		T::initialize_class();
```
Comment (AThousandShips) [inline core/object/class_db.h:218]:
```suggestion
		static_assert(TypesAreSame<typename T::self_type, T>::value, "Class not declared properly, please use GDCLASS.");
		T::initialize_class();
```
Same below, see #81020

File: core/object/class_db.h [hunk 3]
```diff
@@ -185,33 +185,63 @@ class ClassDB {
 	}
 
 	template <class T>
-	static void register_class(bool p_virtual = false) {
+	static void register_class(bool p_virtual = false, bool p_exposed = true) {
```
Comment (dsnopek) [inline core/object/class_db.h:188]: I'm not sure we want/need to add `p_exposed` on the existing `ClassDB::register_*()` functions?

At the moment, unexposed classes actually don't need to get manually registered all, they'll get added to `ClassDB` when the first instance of one of them is created. However, if we do want the ability to manually register them, I think `ClassDB::register_internal_class()` and `GDREGISTER_INTERNAL_CLASS()` would be sufficient?

File: core/object/class_db.h [hunk 4]
```diff
@@ -185,33 +185,63 @@ class ClassDB {
...
 		ERR_FAIL_COND(!t);
-		t->exposed = true;
+		t->exposed = p_exposed;
 		t->class_ptr = T::get_class_ptr_static();
 		t->api = current_api;
 		//nothing
 	}
 
+	template <class T>
+	static void register_virtual_class(bool p_exposed = true) {
```
Comment (dsnopek) [inline core/object/class_db.h:216]: I don't think we need `ClassDB::register_virtual_class()` for anything? You can already register virtual classes via `ClassDB::register_class()` (used by `GDREGISTER_VIRTUAL_CLASS()`).

**Review by dsnopek — COMMENTED** [review]
Thanks! The changes for GDExtension look great.

I'm not so sure about the other changes to `class_db.h`, though. None of those are actually needed for GDExtension (only the changes to `class_db.cpp`), and may require wider discussion because it affects a core Godot API.

That said, I do personally like the addition of `ClassDB::register_internal_class()` and `GDREGISTER_INTERNAL_CLASS()` because we'll need something like that on godot-cpp side, and this way we can give godot-cpp the same API as Godot. But it may be best to leave the rest out (or at least save it for a different PR so the changes we need for GDExtension aren't delayed by discussion of the core API changes).

**Comment by Daylily-Zeleen** [comment]
@dsnopek To be honest, I think the changes your mention about are not needed, too. 
However, `is_virtual`, `is_abstract` and `is_exposed` are in the same struct, I guess they need to be able to control at the same level.

And there already have `register_class(bool p_virtual)` and have `register_abstract_class()` to register a abstract class, therefore I add `register_virtual_class()` and `register_internal_class()`, and add argument `p_exposed` and `p_virtual` to each `register_xxx_class()` if they need.

I think the best way is remove `register_abstract_class()`, `register_virtual_class()` and `register_internal_class()`, pass `p_virtual`, `p_abstract` and `p_exposed` to `register_class()`.
Of course, `GDREGISTER_CLASS()`, `GDREGISTER_VIRTUAL_CLASS()`, `GDREGISTER_ABSTRACT_CLASS()` and `GDREGISTER_INTERNAL_CLASS()` are still needed and base on `register_class()`.

But I guss that `register_abstract_class()` is already in use both godot and godot-cpp, I think this idea can't be realized.
----------------------------------------------

Let's return to this pr, do you means that only provide `ClassDB::register_internal_class()` without argument and `GDREGISTER_INTERNAL_CLASS(m_class)` is enough?

**Comment by dsnopek** [comment]
> Let's return to this pr, do you means that only provide `ClassDB::register_internal_class()` without argument and `GDREGISTER_INTERNAL_CLASS(m_class)` is enough?

Yes, in my opnion, that would be enough.

**Comment by Daylily-Zeleen** [comment]
@dsnopek OK, change again both this pr and [godot-cpp/pull/970](https://github.com/godotengine/godot-cpp/pull/970).

**Review by dsnopek — APPROVED** [review]
Thanks, this looks great to me!

I tested this PR together with PR https://github.com/godotengine/godot-cpp/pull/970 and my SG Physics 2D extension, by registering the editor plugins with `GDREGISTER_INTERNAL_CLASS()`: the editor plugins still worked just fine (meaning they are registered correctly in `ClassDB`), but they were no longer listed as options on the "Create New Node" dialog. So, it appears to be working great as well!

File: core/extension/gdextension.cpp [hunk 5]
```diff
@@ -286,6 +286,7 @@ void GDExtension::_register_extension_class(GDExtensionClassLibraryPtr p_library
 	const GDExtensionClassCreationInfo2 class_info2 = {
 		p_extension_funcs->is_virtual, // GDExtensionBool is_virtual;
 		p_extension_funcs->is_abstract, // GDExtensionBool is_abstract;
+		true,
```
Comment (akien-mga) [inline core/extension/gdextension.cpp:289]:
```suggestion
		true, // GDExtensionBool is_exposed;
```
for consistency with the rest.

**Comment by Daylily-Zeleen** [comment]
@akien-mga Done.

**Comment by dsnopek** [comment]
For some reason this PR likes to _sometimes_ fail CI in the "Linux / Editor with doubles and GCC sanitizers" job with this error:

```
AddressSanitizer: attempting free on address which was not malloc()-ed:
```

The stack trace appears to point to freeing a `PropertyInfo` from a `MethodInfo`'s arguments when Godot is shutting down. But I don't really see how this PR could lead to this issue?

If I re-run the tests enough times, it does eventually pass, but it fails often enough that I'm questioning if there's something I'm missing?

**Comment by AThousandShips** [comment]
It's occuring generally even for completely trivial PRs

**Comment by dsnopek** [comment]
@reduz This adds the new `is_exposed` field to the `GDExtensionClassCreationInfo2` struct (with a `2` at the end) which is accepted by the `classdb_register_extension_class2` function (also with a `2`) which are both new for Godot 4.2. GDExtensions created for Godot 4.1 or earlier use the `GDExtensionClassCreationInfo` struct and `classdb_register_extension_class` function (without a number at the end of either) which doesn't have this field, and will continue working as they used to.

The `GDExtensionClassCreationInfo2` struct and `classdb_register_extension_class2` function were added in PR https://github.com/godotengine/godot/pull/78634, and now other PRs (like this one) can continue to add things to them until Godot 4.2 is released, and after that they'll have to remain as they are for compatibility. There's a couple more PRs that need to change this new struct (including the "hot reload" PR!) and I'm hoping we can get as many as possible merged for 4.2 to avoid needing to do a `3` version of the struct right away for 4.3 :-)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**